Completion Suggester V2

areek opened the pull request:
# Overview
The motivation behind the new completion field is to support auto-complete/search-as-you-type functionality. This is a navigational feature to guide users to relevant results as they are typing, improving search precision. Ideally, auto-complete functionality should be as fast as a user types to provide instant feedback relevant to what a user has already typed in.

This PR introduces a new completion field (https://github.com/elastic/elasticsearch/issues/10746) using Lucene’s new suggest API. The new field has a superset of functionalities provided by the existing completion field (CompletionSuggester and ContextSuggester), namely near-real time search, document retrieval, support for multiple contexts and flexible scoring through index and query-time boosts.

# Use Case

Guide users to relevant documents by suggesting song titles as they type. In this example, the suggestions to serve the user will be indexed as a completion field named `title_suggest`.

The mapping snippet below adds a completion field named `title_suggest`
```json
...
"properties": {
 "title_suggest": {
  "type": "completion"
 }
}
```
To add title suggestions, index documents with `title_suggest` field.
```json
{
 "title_suggest" : "Californication",
}
```
You can also specify an array or an object to specify multiple values and/or configure index-time weights for completion values. **Index-time Weights** (maybe call this index-time boost?) are for sorting suggestions that match a prefix. For example, is "Californication" or "Can't Stop" more relevant for a prefix query of "ca"?

### Querying
To query against a CompletionField, use `_suggest` API (see https://github.com/elastic/elasticsearch/issues/10746 for details). 

Apart from simple **Prefix Query**, you can use **Fuzzy Prefix Query** to get typo-tolerant suggestions. This queries a set of similar prefixes (based on edit distance) and scores the suggestions based on how similar they are to what the user has typed in. For example, a prefix query of "cale" would suggest "Californication" and "Can't Stop" in order. 

There is support for expressing a query prefix as a **Regular Expression**. For example, `ca[l-n]` will suggest titles starting with "cal", "cam" or "can".

### Scoring and Filtering
It is often desirable to serve suggestions filtered and/or boosted by some criteria. For example, you want to suggest song titles filtered by certain artists or you want to boost song titles based on their genre. 

To achieve suggestion filtering and/or boosting, you can add **contexts** while configuring a completion field. You can define multiple contexts for a completion field (NOTE: adding contexts increases the index size). Every context has a unique name and a type. Currently, there are two types `category` and `geo`.

To use genre as a context to song suggestion, a completion field is configured with a 'genre_context' context which indexes values from field 'genre' as shown below
```json
...
"properties": {
 "title_suggest": {
  "type": "completion",
  "contexts" : [
   { "name": "genre_context", "type" : "category", "path": "genre" }
  ]
 }
}
```
Suggestions with contexts can be added by indexing a document as follows:
```json
{
 "title_suggest" : "Californication",
 "genre" : ["funk rock", "alterative_rock", "funk metal"],
}
```
The default behaviour for context-enabled fields is to suggest from *all* contexts. You can filter on defined context(s) at query time by adding query contexts. For example, to restrict song suggestions to only 'funk rock' and 'indie' genres.
```json
"suggest-namespace" : {
 ...
 "completion" : {
  "field" : "title_suggest",
  "contexts": {           
   "genre_context": ["funk rock", "indie"]
  }
 }
}
```
You can boost suggestions for some genres more than others. For example, restrict song suggestions to all genres that start with "funk" but boost song suggestions with "funk rock" genre.
```json
...
"genre_context": [ 
 { "context": "funk", "prefix": true }, 
 { "context": "funk rock", "boost": 4 } 
]
```
Configuring a completion field with multiple contexts, allows filtering/boosting suggestions by multiple criteria. For example, adding artist and genre contexts as follows:
```json
...
"title_suggest": {
 "type": "completion",
 "contexts" : [
  { "name": "genre_context", "type" : "category", "path": "genre" },
  { "name": "artist_context", "type" : "category", "path": "artist" }
 ]
}
```
You can restrict song suggestions to "funk rock" and "alternative rock" genre and boost songs by "rhcp" by adding the following query contexts.
```json
"contexts": {           
 "genre_context": [ "funk rock", "alternative rock" ],
 "artist_context" : { "context" : "rhcp", "boost": 4 }
}
```
The query response for any context-enabled field, will include all the associated context values for that entry, along with the suggestion and weight.

NOTE: Adding more contexts to a completion field will result in search performance degradation (when using match all context) and increase the index size.

### PR status
 - This PR excludes the Lucene related changes ([LUCENE-6459](https://issues.apache.org/jira/browse/LUCENE-6459) and smaller changes) to keep the PR es specific.
 - Retrieving documents with suggestions are not implemented in this PR, and will be done in a subsequent PR
 - Todo: Benchmark context suggester

rjernst commented:
For backcompat, why can't the field type name stay the same, and the logic to decide which actual mapper/fieldtype to create be when parsing (where we have the index created version)?

areek commented:
@rjernst I think that might be confusing to the users, you have the same field type that behaves differently w.r.t. when they were created? 
My thoughts, we use the index created version, to map the the old field type from 'completion' -> 'completion_old' for older versions, and make it clear what is the expected behaviour. If you want to use the new type, you explicitly create a 'completion' field type or if you still rely on features such as attaching payloads (not supported for the new field type) you continue using 'completion_old'. Thoughts?

rjernst commented:
I think that is no more confusing that what I suggested since someone that doesn't know about the new type will get the "new" behavior (they will get a new completion field when they create another with type `completion`).

rjernst commented:
And I don't think we should allow creating the old type.

areek commented:
I agree it being somewhat confusing, it can be solved through documentation (same can be done for your suggestion :)). To me, it is easier to reason about new behaviour tied to a different type name then that of the old? But if we don't allow creating the old type, what you are suggesting makes sense to me. 
So, we will check the index creation version and choose which field-type/suggester to be used. I will explore this, thanks for the suggestion, @rjernst!

clintongormley commented:
The bwc decision is hard. I agree with @rjernst that we shouldn't allow users to create new indices with the old suggester, but that existing indices should still work.

We can choose which suggester to use a query time based on index creation date/version.  The behaviour will be different (e.g. with the old you get back payloads, with the new you get back documents).  I think that almost everybody would use a suggester on a single index, rather than across  multiple indices, so mixed behaviour probably isn't a problem (although perhaps we should not allow combining results from old and new indices?)

If we leave the field type the same (ie `completion`) there is no visual indication to the user that queries will behave differently, but I don't much like the `completion_old` name either...

clintongormley commented:
@areek what is `exact`?  does it mean "not fuzzy"?

areek commented:
@clintongormley when `exact` is false for query context, the context in question is a prefix of some indexed contexts.
For example, if you have indexed contexts `type1`, `type2` and `type3`, you could boost/filter by all three contexts by using a query context of 
```
{ "value": "type", "exact": false, "boost": .. }
```
instead of:
```
{ "value": "type1", "boost": .. }
{ "value": "type2", "boost": .. }
{ "value": "type3", "boost": .. }
```
Maybe we could have a better name for this option? thoughts?

clintongormley commented:
> Maybe we could have a better name for this option? thoughts?

@areek maybe we use `context` instead of `value`, and use `prefix` when it isn't exact?

abhijitiitr commented:
@areek I tested these changes and it feels wonderful to have such a great deal of flexibility. But I faced a few issues. As @clintongormley said
> (e.g. with the old you get back payloads, with the new you get back documents).

I couldn't find a way to get documents currently. If there are multiple inputs having multiple contexts and weights per doc for a `completion` type field, the suggester currently doesn't differentiate between inputs of same and different docs. 
Can you suggest ways of differentiating output on the basis of their docs with these changes?

areek commented:
@abhijitiitr Thanks for testing the new suggester. It would be awesome if you could share any details about the data set you tested it with :)

> I couldn't find a way to get documents currently.

This feature is not implemented in this PR yet, but will be done as a subsequent PR

>  If there are multiple inputs having multiple contexts and weights per doc for a completion type field, the suggester currently doesn't differentiate between inputs of same and different docs. 

This is a known issue and will be fixed. The suggested inputs that correspond to the same doc will be deduplicated.

> Can you suggest ways of differentiating output on the basis of their docs with these changes?

Once the document fetch phase is implemented, the associated document will be returned with the suggestion.

lenniboy commented:
I'm not entirely sure from reading this PR whether it will be possible but it seems that payloads are no longer supported in the new suggester. This would make me quite sad since I discovered a bug regarding payloads in the old suggester and was hoping it would be fixed with the new.

On the other hand I read the following comment: "(e.g. with the old you get back payloads, with the new you get back documents)". This would indeed solve my problem. Is it true that the new suggest API returns the complete document?

areek commented:
> Is it true that the new suggest API returns the complete document?

@lenniboy Yes, so instead of storing a value as a payload, you can store it as a field in the document. You will get back the document with the suggestions.

clintongormley commented:
@areek @jpountz came up with a simple solution to the bwc issue.  Call the new field type and suggester `complete` instead of `completion`.  Problem solved

rjernst commented:
That sounds like a horrible idea. Those two names are far too close to each other: *very* confusing. Why is it a problem to force the new field type on 2.0+ indexes?

clintongormley commented:
@rjernst more confusing than having the same name return different results, and no easy way to distinguish them, and having an exception thrown if you try to do a completion request across old and new indices?  And what about if you try to add a completion field to an old index - what happens then?

rjernst commented:
> @rjernst more confusing than having the same name return different results, and no easy way to distinguish them, and having an exception thrown if you try to do a completion request across old and new indices?

Yes.

> And what about if you try to add a completion field to an old index - what happens then?

The same thing that happens for other mapper changes we've done. The behavior stays the same on indexes before 2.0 (for new fields as well) and the new behavior happens with indexes created on or after 2.0.

clintongormley commented:
> The same thing that happens for other mapper changes we've done. The behavior stays the same on indexes before 2.0 (for new fields as well) and the new behavior happens with indexes created on or after 2.0.

Today you can run a completion suggester across multiple indices. If you try this with what you are proposing, then you'll either (a) not be able to reduce the results or (b) get a mixed result set, some with payloads and some with docs.

Plus the new suggester supports things that the old suggester doesn't, which could lead to more confusing exceptions.  What I don't like about keeping the same name is that it is hard for the user to know which index is going to behave in what way.  Changing the name resolves that: "no field of type `complete` found in index foo"

areek commented:
Updated PR:
 - Backwards compatibility
   - `completion` field created on a pre 2.0 index will use the old implementation
   - `completion` field created on a post 2.0 index will use the new implementation
   - if new and old `completion` field is queried at the same time (e.g. same field name across different indices), then an IAE is thrown at the reduce phase
 - Old Completion suggester now resides in `org.elasticsearch.search.suggest.completion.old`
 - First draft of API docs for new suggester: [Completion Suggester](https://github.com/areek/elasticsearch/blob/comp_suggester_v2/docs/reference/search/suggesters/completion-suggest.asciidoc), [Context Suggester] (https://github.com/areek/elasticsearch/blob/comp_suggester_v2/docs/reference/search/suggesters/context-suggest.asciidoc)
 - Increased test coverage
 - `geo` context mapping work

@clintongormley would be awesome to get some feedback on the API docs (linked above).
@rjernst @jpountz would be great to get some feedback on the new completion field

rjernst commented on core/src/main/java/org/elasticsearch/index/mapper/core/CompletionFieldMapper.java:
```diff
@@ -18,151 +18,98 @@
     public static class Fields {
         // Mapping field names
-        public static final String ANALYZER = "analyzer";
+        public static final ParseField INDEX_ANALYZER = new ParseField("index_analyzer", "analyzer");
```
This should stay just `analyzer`. Behavior was changed for all field mappers to have just `analyzer`, and `search_analyzer` simply overrides the search time behavior.

abhijitiitr commented:
@areek Sorry for the late response. We have around 0.5M restaurants for which we wanted to index the dishes they served which could be scoped under different groups and scored differently.
```javascript
 [
        {
            "input" : ["Palak Paneer", "Shahi Paneer"],
            "contexts" : {
                "cuisine_type" : ["North Indian","Continental"],
                "meal_type" : ["Dinner","Lunch"]
            },
            "weight" : 31
        },
       {
            "input" : ["Upma"],
            "contexts" : {
                "cuisine_type" : ["South Indian","Continental"],
                "meal_type" : ["Breakfast"]
            },
            "weight" : 29
        },
        {
            "input" : ["Pizza", "Rissoto", "Pasta"],
            "contexts" : {
                "cuisine_type" : ["Italian"],
                "meal_type" : ["Lunch","Dinner"]
            },
            "weight" : 34
        }
        
    ]
```
An example of an input to the suggester is given above. 

Regarding the document linking part, I was studying the codebase and was able to find the trace of new suggest functionality.
The `innerExecute` function in the `CompletionSuggester` class was an intermediary in bringing the suggestions and we need to link these suggestions to the documents . As you had earlier said 
> Once the document fetch phase is implemented, the associated document will be returned with the suggestion.

I'm not being able to find a way to link the documents linked to the `outputs`(i.e `options` which are finally added in the end of the function. I only know the codebase in bits and pieces and would be very grateful for your help.

clintongormley commented on docs/reference/search/suggesters/completion-suggest.asciidoc:
```diff
@@ -79,22 +68,20 @@ Mapping supports the following parameters:
     characters per input string in order to prevent massive inputs from
     bloating the underlying datastructure. The most usecases won't be influenced
```
The most -> Most

clintongormley commented on docs/reference/search/suggesters/completion-suggest.asciidoc:
```diff
@@ -79,22 +68,20 @@ Mapping supports the following parameters:
     characters per input string in order to prevent massive inputs from
     bloating the underlying datastructure. The most usecases won't be influenced
     by the default value since prefix completions hardly grow beyond prefixes longer
```
hardly -> seldom

clintongormley commented on docs/reference/search/suggesters/completion-suggest.asciidoc:
```diff
@@ -238,3 +209,40 @@ The following parameters are supported:
 NOTE: If you want to stick with the default values, but
```
On `unicode_aware`, just above this say:

    If `true`, all measurements (like fuzzy edit distance, transpositions, and lengths) are measured in Unicode 
    code points instead of in bytes.  This is slightly slower than raw bytes, so it is set to `false` by default.

clintongormley commented on docs/reference/search/suggesters/completion-suggest.asciidoc:
```diff
@@ -238,3 +209,40 @@ The following parameters are supported:
+    Possible flags are `ALL` (default), `ANYSTRING`, `COMPLEMENT`,
+    `EMPTY`, `INTERSECTION`, `INTERVAL`, or `NONE`. Please check the
+    http://lucene.apache.org/core/4_9_0/core/org/apache/lucene/util/automaton/RegExp.html[Lucene
+    documentation] for their meaning
```
You can link to <<regexp-syntax>> instead.

clintongormley commented on docs/reference/search/suggesters/context-suggest.asciidoc:
```diff
@@ -1,237 +1,269 @@
+To achieve suggestion filtering and/or boosting, you can add context mappings while
+configuring a completion field. You can define multiple context mappings for a
+completion field.
+Every context mapping has a unique name and a type. Currently, there are two types
```
there are two types: `category` and `geo`

clintongormley commented on docs/reference/search/suggesters/context-suggest.asciidoc:
```diff
@@ -1,237 +1,269 @@
+configuring a completion field. You can define multiple context mappings for a
+completion field.
+Every context mapping has a unique name and a type. Currently, there are two types
+*category* and *geo*. The context mappings is configured under the `contexts` parameter
```
The context mappings ->

    Context mappings are configured with the `contexts` parameter in the field mapping.

clintongormley commented on docs/reference/search/suggesters/context-suggest.asciidoc:
```diff
@@ -1,237 +1,269 @@
-is used, when ever no specific is provided for the certain context. Note that a context is
-defined by at least one value. The `type` option defines the kind of information hold by this
-context. These type will be explained further in the following sections.
+NOTE: Adding context mappings increases the index size for completion field.
```
Is it possible to give some idea of how much more memory is used based on the number of contexts and the cardinality?

clintongormley replied:
I'd at least add a link to the completion stats, which people can use to check how much memory is being used.

areek replied:
linked to completion stats for now, I am working on getting a estimate for memory usage for contexts..

clintongormley commented on docs/reference/search/suggesters/context-suggest.asciidoc:
```diff
@@ -1,237 +1,269 @@
-specify this field of the documents that should be used. If the referenced field contains multiple
-values, all these values will be used as alternative categories.
+
+The `category` context allows you to associate one or more categories to suggestions at index time.
```
to suggestions -> with suggestions

clintongormley commented on docs/reference/search/suggesters/context-suggest.asciidoc:
```diff
@@ -1,237 +1,269 @@
+    {
+        "name": "cat_context",
         "type": "category",
-        "default": ["red", "orange"]
```
we no longer support default values for contexts?

areek replied:
No we don't, but only because we can query across all contexts (but will be expensive). Maybe we should support optional default value for contexts, need to give this a bit more thought

clintongormley replied:
I think the default value is (also?) used at index time, so if no context is specified, then it gets the default context.  Not sure how useful that is in practice.

clintongormley commented on docs/reference/search/suggesters/context-suggest.asciidoc:
```diff
@@ -1,237 +1,269 @@
+[float]
+===== Indexing category contexts
+
+`category` contexts can be explicitly set with suggestions or be indexed from another field in the
```
Category contexts can be specified explicitly when indexing suggestions, or can be retrieved from another 
    field in the document via the `path` parameters. If a suggestion has multiple contexts (categories?),
    the suggestion will be indexed for each category:  (followed by the example)

clintongormley commented on docs/reference/search/suggesters/context-suggest.asciidoc:
```diff
@@ -1,237 +1,269 @@
+    "suggest": {
+        "input": ["timmy's", "starbucks", "dunkin donuts"],
+        "context": {
+            "place_type": ["cafe", "food"]
```
"place_type": ["cafe", "food"] <1>

clintongormley commented on docs/reference/search/suggesters/context-suggest.asciidoc:
```diff
@@ -1,237 +1,269 @@
     }
 }
 --------------------------------------------------
 
```
<1> This suggestion will be associated with each specified category.

clintongormley commented on docs/reference/search/suggesters/context-suggest.asciidoc:
```diff
@@ -1,237 +1,269 @@
+of contexts.
+
+`category` contexts can also be referenced from another indexed field in the document. The following
+will index the values of the 'cat' field, if the context mapping was configured with the `path`
```
I'd provide an example of the mapping followed by the indexing request. Possibly also specify an explicit category, and add a callout `<1>` to say "This suggestion will be indexed with the following categories: [ explicit, cafe, food ]

clintongormley commented on docs/reference/search/suggesters/context-suggest.asciidoc:
```diff
@@ -1,237 +1,269 @@
-of a specified geolocation. At index time, a lat/long geo point is converted into a
-geohash of a certain precision, which provides the context.
+
+A `geo` context allows you to associate one or more geo point or geohash with suggestions
```
geo pointS or geohashES

clintongormley commented on docs/reference/search/suggesters/context-suggest.asciidoc:
```diff
@@ -1,237 +1,269 @@
-The mapping for a geo context accepts four settings, only of which `precision` is required:
+===== Geo Mapping
+
+In addition to `path` setting, `geo` context mapping accepts the following settings:
```
In addition to THE path...

clintongormley commented on docs/reference/search/suggesters/context-suggest.asciidoc:
```diff
@@ -1,237 +1,269 @@
+`precision`::
+    This defines the precision of the geohash to be indexed and can be specified
+    as a distance value (`5m`, `10km` etc.), or as a raw geohash precision (`1`..`12`).
+    Defaults to a raw geohash precision value of `12`.
```
Is this the right default?  Won't it make for very big FSTs? It amounts to centimetre level, which seems way over the top.  I'd think km level (6) would be a better default

clintongormley replied:
See https://www.elastic.co/guide/en/elasticsearch/guide/current/geohashes.html

areek replied:
Agreed, precision of 6 is a better default, changed

clintongormley commented on docs/reference/search/suggesters/context-suggest.asciidoc:
```diff
@@ -239,24 +271,22 @@ PUT services/service/1
-simple case all precision values defined in the mapping will be applied to the given
-location.
+Suggestions can be filtered and boosted with respect to how close they are to one or
+more geo points. The following filters suggestions that fall within geo point:
```
that fall within geo point?  Not sure what that means.  Surely you always want distance involved here, as just a 0.0000001 difference in a point would not return the document?

clintongormley replied:
OK so you're using geohashes here - worth explaining how a geopoint gets translated internally.

clintongormley commented on docs/reference/search/suggesters/context-suggest.asciidoc:
```diff
@@ -264,54 +294,63 @@ POST services/_suggest
+                        "lon": -79.380
                     },
-                    "precision": "1km"
-                }
```
Precision has been removed?  Does that make sense?

areek replied:
I added back precision setting for geo query contexts.

clintongormley commented:
@areek looking good. i've added some minor comments and some questions.  One thing about the docs: category and context are used interchangeably, which i found quite confusing. I'd try to define the difference and then them consistently.

areek commented:
@clintongormley Thanks for the review, I updated the PR incorporating your feedback.

clintongormley commented:
@areek LGTM from a docs/api point of view.  I'd get a developer review as well.

thanks!

mikemccand commented on core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionFieldStats.java:
```diff
@@ -0,0 +1,62 @@
+                            continue;
+                        }
+                        if (Regex.simpleMatch(fields, fieldName)) {
+                            long fstSize = completionTerms.ramBytesUsed();
```
Maybe do this above so `completionTerms.ramBytesUsed()` is only called once?

areek replied:
Thanks for catching that, I re-structured the method to call `completionTerms.ramBytesUsed()` only once.

mikemccand commented on core/src/main/java/org/elasticsearch/search/suggest/completion/context/ContextMappings.java:
```diff
@@ -0,0 +1,322 @@
+            contexts.addAll(mapping.parseContext(document));
+            final Iterator<CharSequence> contextsIterator = contexts.iterator();
+            final CharsRefBuilder scratch = new CharsRefBuilder();
+            scratch.append(((char) typeId));
```
Hmm, how large is typeId allowed to be (how many unique context types are allowed)?  It's an int, but we cast to char here, and javadocs say it's a byte :)  Can we check this and throw a clear exc if user tries to make too many?

areek replied:
IMO, supporting a maximum of `10` unique context types seems reasonable? Now we throw an UnsupportedOperationEx if a user tries to define more than 10 unique context type per field

mikemccand replied:
10 seems a bit arbitrary?  But yeah it "seems" high enough :)  I think I would just limit it to the 1 byte max (255?)...

areek replied:
Changed the max unique contexts to `255`

mikemccand commented on docs/reference/search/suggesters/context-suggest.asciidoc:
```diff
@@ -1,237 +1,290 @@
-is used, when ever no specific is provided for the certain context. Note that a context is
-defined by at least one value. The `type` option defines the kind of information hold by this
-context. These type will be explained further in the following sections.
+NOTE: Adding context mappings increases the index size for completion field. You can monitor
```
Can you add that it also increases the heap usage?  Ie the FST is heap resident...

areek replied:
I updated the docs to clarify that the completion index is head resident

mikemccand commented on core/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -655,15 +656,13 @@ public SuggestStats suggestStats() {
-        try {
+        try (final Engine.Searcher currentSearcher = acquireSearcher("completion_stats")) {
             PostingsFormat postingsFormat = PostingsFormat.forName(Completion090PostingsFormat.CODEC_NAME);
             if (postingsFormat instanceof Completion090PostingsFormat) {
```
Isn't this always true (pre-existing)?

areek replied:
It is always true, I removed the check

mikemccand commented on core/src/main/java/org/elasticsearch/index/shard/IndexShard.java:
```diff
@@ -655,15 +656,13 @@ public SuggestStats suggestStats() {
             }
-        } finally {
-            currentSearcher.close();
+            completionStats.add(CompletionFieldStats.completionStats(currentSearcher.reader(), fields));
```
Can we do this inside the `try` statement and remove the finally clause?

areek replied:
Not sure I understand, before we had the finally clause but with the change I removed the finally clause and used try-with-resources on the engine.searcher?

mikemccand commented on core/src/main/java/org/elasticsearch/search/suggest/completion/CompletionSuggester.java:
```diff
@@ -53,68 +58,99 @@
+        TopSuggestDocsCollector collector = new TopSuggestDocsCollector(suggestionContext.getSize());
+        suggest(searcher, toQuery(suggestionContext), collector);
+        for (TopSuggestDocs.SuggestScoreDoc suggestDoc : collector.get().scoreLookupDocs()) {
+            // TODO: currently we can get multiple entries with the same docID
```
Is there a Lucene issue open to do this dedup'ing in {{TopNSearcher}}?

areek replied:
No not yet, but I am working on this one. Dedup'ing entries with early termination in the `TopSuggestDocsCollector` seems tricky. One way is to terminate as soon as we collect `num` unique doc ids from the `TopNSearcher` but still collect `num * maxEntriesPerDoc` entries in the priority queue. On `TopSuggestDocsCollector.get` merge the entries with the same doc id and present only `num` `SuggestScoreDocs`. Any thoughts on this?

mikemccand replied:
This is indeed tricky.  To always be correct in a single pass, yeah, we must keep up to `num * maxEntriesPerDoc` entries in the queue ... we don't know the docID of a given path until the path finishes, so we have to dedup in collection not as we enumerate partial paths ...

mikemccand commented:
I left several small comments but overall this looks really nice.

It's a big change, I think we need to get it in soon so it has more time to bake before 2.0 GA (it's already scary close, and it's weird that beta1 won't have it but GA will), and then make new PRs for further additions (e.g. retrieving documents sources/fields for each suggestion, generic filters).

Do we talk about the changes vs the current suggesters in the migration guide?  Or maybe we should do a blog post about the differences?

areek commented:
Thanks @mikemccand for the review. I have addressed all your feedback. I will create a Lucene issue for the deduping doc ids in TopSuggestDocsCollector and back-port the fix.

mikemccand commented:
Thanks @areek, LGTM.

areek commented:
This PR has been merged to a feature branch https://github.com/elastic/elasticsearch/tree/completion_suggester_v2

jrots commented:
When will this be available in v2.0?
